test/gtest/uct/uct_test.cc

```diff
@@ -26,6 +26,66 @@ std::string resource::name() const {
     return ss.str();
 }
 
+resource::resource() : md_name(""), tl_name(""), dev_name(""),
+                       dev_type(UCT_DEVICE_TYPE_LAST)
+{
+    CPU_ZERO(&local_cpus);
+    memset(&listen_if_addr, 0, sizeof(listen_if_addr));
+    memset(&connect_if_addr, 0, sizeof(connect_if_addr));
+}
+
+resource::resource(const std::string& _md_name, const cpu_set_t& _local_cpus,
+                   const std::string& _tl_name, const std::string& _dev_name,
+                   uct_device_type_t _dev_type) :
+                   md_name(_md_name), local_cpus(_local_cpus), tl_name(_tl_name),
+                   dev_name(_dev_name), dev_type(_dev_type)
+{
+    memset(&listen_if_addr, 0, sizeof(listen_if_addr));
+    memset(&connect_if_addr, 0, sizeof(connect_if_addr));
+}
+
+resource_speed::resource_speed(const uct_md_h& md, const std::string& _md_name,
+                               const cpu_set_t& _local_cpus, const std::string& _tl_name,
+                               const std::string& _dev_name, uct_device_type_t _dev_type) :
+                               resource(_md_name, _local_cpus, _tl_name, _dev_name, _dev_type) {
+    ucs_status_t status;
+    ucs_async_context_t *async;
+    uct_worker_h worker;
+    uct_iface_params_t iface_params = { 0 };
+    uct_iface_config_t *iface_config;
+    uct_iface_attr_t iface_attr;
+    uct_iface_h iface;
+
+    status = ucs_async_context_create(UCS_ASYNC_MODE_THREAD_SPINLOCK, &async);
+    ASSERT_UCS_OK(status);
+
+    status = uct_worker_create(async, UCS_THREAD_MODE_SINGLE, &worker);
+    ASSERT_UCS_OK(status);
+
+    status = uct_md_iface_config_read(md, tl_name.c_str(), NULL, NULL, &iface_config);
+    ASSERT_UCS_OK(status);
+
+
+    iface_params.field_mask           = UCT_IFACE_PARAM_FIELD_OPEN_MODE |
+        UCT_IFACE_PARAM_FIELD_DEVICE;
+    iface_params.open_mode            = UCT_IFACE_OPEN_MODE_DEVICE;
+    iface_params.mode.device.tl_name  = tl_name.c_str();
+    iface_params.mode.device.dev_name = dev_name.c_str();
+
+    status = uct_iface_open(md, worker, &iface_params, iface_config, &iface);
+    ASSERT_UCS_OK(status);
+
+    status = uct_iface_query(iface, &iface_attr);
+    ASSERT_UCS_OK(status);
+
+    bw = iface_attr.bandwidth;
+
+    uct_iface_close(iface);
+    uct_config_release(iface_config);
+    uct_worker_destroy(worker);
+    ucs_async_context_destroy(async);
```

Comment:
maybe we won't create worker and async context for every resources? it could be long operation since it may create and destroy an async thread every time

Reply:
moved the async and worker objects allocation/release to `uct_test::enum_resources`
the worker object is passed to the `resource_speed` ctor